python/ray/serve/_private/http_state.py

```diff
@@ -404,3 +429,17 @@ async def ensure_http_route_exists(self, endpoint: EndpointTag, timeout_s: float
                 for proxy in self._proxy_states.values()
             ]
         )
+
+    def update_active_flags(self, active_nodes: Set[str]):
+        """Update the active states of all HTTP proxies.
+
+        Given a set of active nodes, set the active flag of all HTTP proxies, except
+        for head node. Head node will always be active.
+        """
+        for node_id, proxy_state in self._proxy_states.items():
+            # Head node will always be active.
+            if node_id == self._head_node_id:
```

Comment:
Does `self._head_node_id` get updated whenever there's a new head node?

Reply:
This is out of scope of this issue. But I don't see code to update `self._head_node_id`. My guess is if there is a new head node, the head node will create new controller and new http state so this will no longer be used. The controller on the new head node should determine the state of new http proxies as well 🙂

Reply:
Ah, yeah that makes sense.